File: functions/consulta.mjs
import * as cheerio from "cheerio";
import { sesionValida } from "./login.mjs";
import { ipBloqueada, limiteConsulta, ocultarConsulta, pausarConsultas, registrarEvento } from "./security.mjs";

const BASE_URL = "https://mi.nosis.com";

function jsonResponse(status, payload) {
  return new Response(JSON.stringify(payload), {
    status,
    headers: {
      "Content-Type": "application/json; charset=utf-8",
    },
  });
}

function getCookieHeader(sessionId, nstk) {
  return `ASP.NET_SessionId=${sessionId}; nstk=${nstk}`;
}

export default async function handler(request) {
  const ip = request.headers.get("x-nf-client-connection-ip") || "unknown";
  if (await ipBloqueada(ip)) {
    await registrarEvento(request, { type: "blocked_request", outcome: "denied" });
    return jsonResponse(403, { ok: false, error: "Acceso bloqueado." });
  }

  if (!sesionValida(request)) {
    await registrarEvento(request, { type: "unauthorized_query", outcome: "denied" });
    return jsonResponse(401, {
      ok: false,
      error: "Iniciá sesión para usar esta herramienta.",
    });
  }

  if (request.method !== "POST") {
    return jsonResponse(405, { ok: false, error: "Método no permitido." });
  }

  let body;
  try {
    body = await request.json();
  } catch {
    return jsonResponse(400, { ok: false, error: "El cuerpo no es JSON válido." });
  }

  const entrada = String(body.documento || body.query || "").trim();
  const esNumerico = /^[\d\s-]+$/.test(entrada);
  const documento = esNumerico
    ? entrada.replace(/[\s-]/g, "")
    : entrada.replace(/\s+/g, " ");

  if (!documento) {
    await registrarEvento(request, { type: "invalid_query", outcome: "denied" });
    return jsonResponse(400, { ok: false, error: "Ingresá un DNI o CUIT." });
  }

  const sessionId = process.env.ASP_NET_SESSION_ID;
  const nstk = process.env.NSTK;

  if (!sessionId || !nstk) {
    return jsonResponse(500, {
      ok: false,
      error: "Faltan las variables de entorno de la sesión.",
    });
  }

  const limite = await limiteConsulta(ip);
  if (!limite.allowed) {
    await registrarEvento(request, { type: "query_limited", outcome: "denied", reason: limite.reason });
    return jsonResponse(429, {
      ok: false,
      error: limite.reason === "provider_pause"
        ? "Las consultas están pausadas temporalmente porque el proveedor aplicó un límite."
        : limite.reason === "daily_limit"
          ? "Se alcanzó el límite diario de consultas."
          : limite.reason === "storage_unavailable"
            ? "El control de seguridad no está disponible. Intentá más tarde."
          : "Esperá unos segundos antes de volver a consultar.",
      retryAt: limite.retryAt,
    });
  }

  await registrarEvento(request, {
    type: "query",
    outcome: "started",
    query: ocultarConsulta(documento),
    queryKind: esNumerico ? "documento" : "nombre",
  });

  const cookie = getCookieHeader(sessionId, nstk);
  const headersBusqueda = {
    Accept: "application/json, text/javascript, */*; q=0.01",
    "Content-Type": "application/json; charset=UTF-8",
    "X-Requested-With": "XMLHttpRequest",
    Origin: BASE_URL,
    Referer: `${BASE_URL}/InformeTerceros`,
    Cookie: cookie,
  };

  let resultado;
  try {
    const response = await fetch(`${BASE_URL}/InformeTerceros/RealizarBusqueda`, {
      method: "POST",
      headers: headersBusqueda,
      body: JSON.stringify({ query: documento }),
      signal: AbortSignal.timeout(30000),
    });

    if (!response.ok) {
      if (response.status === 403 || response.status === 429) {
        await pausarConsultas(ip);
        await registrarEvento(request, { type: "provider_limited", outcome: "paused", status: response.status });
      }
      throw new Error(`HTTP ${response.status}`);
    }

    resultado = await response.json();
  } catch (error) {
    return jsonResponse(502, {
      ok: false,
      error: `No se pudo consultar Nosis: ${error.message}`,
    });
  }

  if (resultado.Error) {
    return jsonResponse(502, { ok: false, error: String(resultado.Error) });
  }

  if (!resultado.BusquedaRealizada) {
    return jsonResponse(404, {
      ok: false,
      error: "La búsqueda no fue realizada.",
    });
  }

  if (!resultado.HtmlTabla) {
    return jsonResponse(404, {
      ok: false,
      error: "Nosis no devolvió resultados.",
    });
  }

  const $ = cheerio.load(resultado.HtmlTabla);
  const resultados = $(".content-datos-resultados")
    .map((_, bloque) => {
      const bloque$ = $(bloque);
      const nombre =
        bloque$.find(".resultados-razon-social span").first().text().trim() || "-";
      const documentoResultado =
        bloque$.find("#Documento").first().attr("value")?.trim() || "";
      const direcciones = bloque$
        .find(".resultados-lugar .content-direccion .direccion")
        .map((__, elemento) => $(elemento).text().replace(/\s+/g, " ").trim())
        .get()
        .filter(Boolean);

      let provincia = "-";
      let localidad = "-";
      if (direcciones.length > 0) {
        const partes = direcciones[0]
          .split("-")
          .map((parte) => parte.trim())
          .filter(Boolean);
        provincia = partes[0] || "-";
        localidad = partes[1] || "-";
      }

      return {
        nombre,
        documento: documentoResultado,
        provincia,
        localidad,
        direcciones,
      };
    })
    .get();

  if (resultados.length === 0) {
    return jsonResponse(404, {
      ok: false,
      error: "No se encontraron resultados.",
    });
  }

  const seleccionado = resultados[0];
  if (!seleccionado.documento) {
    return jsonResponse(502, {
      ok: false,
      error: "No se encontró el documento interno del resultado.",
    });
  }

  let fechaNacimiento = "-";
  try {
    const responseFecha = await fetch(
      `https://clientes.credicuotas.com.ar/v1/onboarding/resolvecustomers/${encodeURIComponent(seleccionado.documento)}`,
      {
        headers: { Accept: "application/json" },
        signal: AbortSignal.timeout(30000),
      },
    );
    if (responseFecha.ok) {
      const datosFecha = await responseFecha.json();
      const registro = Array.isArray(datosFecha)
        ? (datosFecha[0] || {})
        : (datosFecha || {});
      fechaNacimiento = registro.fechanacimiento
        || registro.fechaNacimiento
        || registro.fecha_nacimiento
        || registro.fechaNac
        || registro.birthDate
        || "-";
      if (fechaNacimiento === "-" && registro.customer) {
        fechaNacimiento = registro.customer.fechanacimiento
          || registro.customer.fechaNacimiento
          || "-";
      }
    }
  } catch {
    // La fecha es opcional; la consulta principal puede continuar.
  }

  return jsonResponse(200, {
    ok: true,
    resultados: resultados.map((persona) => ({
      nombre: persona.nombre,
      documento: persona.documento,
      cuit: persona.documento,
      provincia: persona.provincia,
      localidad: persona.localidad,
      direccion: persona.direcciones.join("\n") || "-",
      fechaNacimiento: persona === seleccionado ? fechaNacimiento : "-",
    })),
    nombre: seleccionado.nombre,
    documento: seleccionado.documento,
    cuit: seleccionado.documento,
    fechaNacimiento,
    provincia: seleccionado.provincia,
    localidad: seleccionado.localidad,
    direccion: seleccionado.direcciones.join("\n") || "-",
  });
}

File: functions/consulta.py
import base64
import json
import os

import requests
from bs4 import BeautifulSoup


BASE_URL = "https://mi.nosis.com"


def respuesta(status_code, payload):
    return {
        "statusCode": status_code,
        "headers": {
            "Content-Type": "application/json; charset=utf-8",
        },
        "body": json.dumps(payload, ensure_ascii=False),
    }


def handler(event, context):
    try:
        body = json.loads(event.get("body") or "{}")
    except (TypeError, ValueError):
        return respuesta(400, {"ok": False, "error": "El cuerpo no es JSON válido."})

    documento = str(body.get("documento") or body.get("query") or "").strip()
    documento = documento.replace(" ", "").replace("-", "")

    if not documento:
        return respuesta(400, {"ok": False, "error": "Ingresá un DNI o CUIT."})

    session_id = os.environ.get("ASP_NET_SESSION_ID")
    nstk = os.environ.get("NSTK")

    if not session_id or not nstk:
        return respuesta(500, {
            "ok": False,
            "error": "Faltan las variables de entorno de la sesión.",
        })

    session = requests.Session()
    session.cookies.set(
        "ASP.NET_SessionId",
        session_id,
        domain="mi.nosis.com",
        path="/",
    )
    session.cookies.set("nstk", nstk, domain=".nosis.com", path="/")

    headers_busqueda = {
        "Accept": "application/json, text/javascript, */*; q=0.01",
        "Content-Type": "application/json; charset=UTF-8",
        "X-Requested-With": "XMLHttpRequest",
        "Origin": BASE_URL,
        "Referer": f"{BASE_URL}/InformeTerceros",
    }

    try:
        response = session.post(
            f"{BASE_URL}/InformeTerceros/RealizarBusqueda",
            headers=headers_busqueda,
            json={"query": documento},
            timeout=30,
        )
        response.raise_for_status()
        resultado = response.json()
    except (requests.RequestException, ValueError) as error:
        return respuesta(502, {
            "ok": False,
            "error": f"No se pudo consultar Nosis: {error}",
        })

    if resultado.get("Error"):
        return respuesta(502, {"ok": False, "error": str(resultado["Error"])})

    if not resultado.get("BusquedaRealizada"):
        return respuesta(404, {"ok": False, "error": "La búsqueda no fue realizada."})

    html = resultado.get("HtmlTabla", "")
    if not html:
        return respuesta(404, {"ok": False, "error": "Nosis no devolvió resultados."})

    soup = BeautifulSoup(html, "html.parser")
    resultados = []

    for bloque in soup.select(".content-datos-resultados"):
        elemento_nombre = bloque.select_one(".resultados-razon-social span")
        elemento_documento = bloque.select_one("#Documento")
        nombre = (
            elemento_nombre.get_text(" ", strip=True)
            if elemento_nombre
            else "-"
        )
        documento_resultado = (
            elemento_documento.get("value", "").strip()
            if elemento_documento
            else ""
        )
        direcciones = []

        for elemento_direccion in bloque.select(
            ".resultados-lugar .content-direccion .direccion"
        ):
            texto = " ".join(elemento_direccion.get_text(" ", strip=True).split())
            if texto:
                direcciones.append(texto)

        provincia = "-"
        localidad = "-"
        if direcciones:
            partes = [
                parte.strip()
                for parte in direcciones[0].split("-")
                if parte.strip()
            ]
            if partes:
                provincia = partes[0]
            if len(partes) >= 2:
                localidad = partes[1]

        resultados.append({
            "nombre": nombre,
            "documento": documento_resultado,
            "provincia": provincia,
            "localidad": localidad,
            "direcciones": direcciones,
        })

    if not resultados:
        return respuesta(404, {"ok": False, "error": "No se encontraron resultados."})

    seleccionado = resultados[0]
    documento_seleccionado = seleccionado["documento"]
    if not documento_seleccionado:
        return respuesta(502, {
            "ok": False,
            "error": "No se encontró el documento interno del resultado.",
        })

    fecha_nacimiento = "-"
    try:
        response_fecha = requests.get(
            "https://clientes.credicuotas.com.ar/"
            f"v1/onboarding/resolvecustomers/{documento_seleccionado}",
            timeout=30,
        )
        response_fecha.raise_for_status()
        datos_fecha = response_fecha.json()
        if isinstance(datos_fecha, list) and datos_fecha:
            fecha_nacimiento = datos_fecha[0].get("fechanacimiento", "-")
    except (requests.RequestException, ValueError):
        pass

    headers_informe = {
        "Accept": "application/json, text/javascript, */*; q=0.01",
        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        "X-Requested-With": "XMLHttpRequest",
        "Origin": BASE_URL,
        "Referer": f"{BASE_URL}/InformeTerceros",
    }

    try:
        response_informe = session.post(
            f"{BASE_URL}/Informes/DescargarInforme",
            headers=headers_informe,
            data={"documento": documento_seleccionado},
            timeout=60,
        )
        response_informe.raise_for_status()
        pdf = (
            "data:application/pdf;base64,"
            + base64.b64encode(response_informe.content).decode("ascii")
        )
    except requests.RequestException:
        pdf = None

    return respuesta(200, {
        "ok": True,
        "nombre": seleccionado["nombre"],
        "documento": documento_seleccionado,
        "cuit": documento_seleccionado,
        "fechaNacimiento": fecha_nacimiento,
        "provincia": seleccionado["provincia"],
        "localidad": seleccionado["localidad"],
        "direccion": "\n".join(seleccionado["direcciones"]) or "-",
        "pdf": pdf,
        "pdfNombre": f"informe-{documento_seleccionado}.pdf",
    })